File: python/ff.cxx
#include "pymod.hxx"
#include <pybind11/stl.h>
#include <msys/ff.hxx>

using namespace desres::msys::ff;

namespace desres { namespace msys {

void export_ff(module m) {

    class_<Rules>(m, "Rules")
        .def(init<>())
        .def_readwrite("vdw_func", &Rules::vdw_func)
        .def_readwrite("vdw_rule", &Rules::vdw_rule)
        .def_readwrite("exclusions", &Rules::exclusions)
        .def_readwrite("es_scale", &Rules::es_scale)
        .def_readwrite("lj_scale", &Rules::lj_scale)
        ;

    class_<Tuples>(m, "Tuples")
        .def(init<>())
        .def("build", [](Tuples& self, SystemPtr mol, IdList const& fragment) { build(self, mol, fragment); })
        ;
    
    class_<Forcefield>(m, "Forcefield")
        .def(init<>())
        .def_readwrite("name", &Forcefield::name)
        .def_readonly("rules", &Forcefield::rules)
        .def("build_component", [](Forcefield const& self, Component p, SystemPtr mol, Tuples const& t) {
            switch(p) {
                case Component::exclusions:
                    build<Component::exclusions>(mol, self, t);
                    break;
                default:
                    MSYS_FAIL("Unsupported component");
            };
            })
        ;

    enum_<Component>(m, "Component")
        .value("exclusions", Component::exclusions)
        ;

}

}}

File: python/ff.py
from . import _msys


class experimental:
    @staticmethod
    def make_rules(rules, flavor):
        if flavor != "amber":
            raise ValueError("unsupported flavor '%s'; must be 'amber'" % flavor)
        rules.vdw_func = "vdw_12_6"
        rules.vdw_rule = "arithmetic/geometric"
        rules.exclusions = 4
        rules.es_scale = [0.0, 0.0, 0.8333]
        rules.lj_scale = [0.0, 0.0, 0.5]
        return rules

    @staticmethod
    def build_pairs_and_exclusions(mol, flavor="amber"):
        """Build all pair_12_6_es and exclusion terms for the given system

        Arguments:
            mol (msys.System): chemical system
            flavor (str): 'amber' for Amber-style exclusion and 1-4 scaling

        Notes:
            Currently does not support systems with pseudoatoms (virtuals or drudes).
            This is checked.

        """
        if mol.selectIds("mass <= 0"):
            raise ValueError("Systems with pseudos not yet supported")

        tuples = _msys.Tuples()
        for frag in mol.updateFragids():
            tuples.build(mol._ptr, [a.id for a in frag])

        ff = _msys.Forcefield()
        experimental.make_rules(ff.rules, flavor)
        ff.build_component(_msys.Component.exclusions, mol._ptr, tuples)
